src/main/java/com/es/phoneshop/model/cart/cartService/DefaultCartService.java

```diff
@@ -51,28 +49,96 @@ public Cart getCart(HttpServletRequest request) {
         synchronized (cart) {
 
-            if (quantity <= 0) {
-                throw new QuantityLessThenZeroException("Quantity must be positive number");
-            }
-
             Product product = productDao.getProduct(productId);
             if (product == null) {
-                throw new ProductNotFoundInDaoException("Product not exist");
-            }
-            if (product.getStock() < quantity) {
-                throw new ProductStockLessThenRequiredException("Stock: " + product.getStock() + " is less then required: " + quantity);
+                throw new IllegalArgumentException("Product not exist");
             }
 
-            CartItem presentItem = cart.getItems().stream().filter(it -> product.equals(it.getProduct())).findAny().orElse(null);
+            CartItem presentItem = cart.getItems().stream()
+                    .filter(item -> productId.equals(item.getProduct().getId()))
+                    .findAny()
+                    .orElse(null);
+
             if (presentItem != null) {
-                presentItem.increaseQuantity(quantity);
+                update(cart, productId, presentItem.getQuantity() + newQuantity);
+
             } else {
-                cart.getItems().add(new CartItem(product, quantity));
+                checkPositive(newQuantity);
+                checkStockEnough(product, newQuantity);
+
+                CartItem newItem = new CartItem(product, newQuantity);
+                cart.getItems().add(newItem);
+                product.setStock(product.getStock() - newQuantity);
+
+                recalculateCart(cart);
+            }
+        }
+    }
+
+    @Override
+    public void update(Cart cart, Long productId, int newQuantity) {
+        synchronized (cart) {
+
+            checkPositive(newQuantity);
+
+            Product product = productDao.getProduct(productId);
+            CartItem presentItem = cart.getItems().stream()
+                    .filter(item -> productId.equals(item.getProduct().getId()))
+                    .findAny()
+                    .orElse(null);
+
+            int oldQuality = presentItem.getQuantity();
+            if (oldQuality == newQuantity) {
+                return;
+            }
+            if (oldQuality < newQuantity) {
+                checkStockEnough(product, newQuantity - oldQuality);
             }
-            product.setStock(product.getStock() - quantity);
+
+            product.setStock(product.getStock() - (newQuantity - oldQuality));
+            presentItem.setQuantity(newQuantity);
+
+            recalculateCart(cart);
         }
     }
+
+    @Override
+    public void delete(Cart cart, Long productId) {
+        CartItem itemToDelete = cart.getItems().stream()
+                .filter(item -> productId.equals(item.getProduct().getId()))
+                .findAny()
+                .orElse(null);
+
+        itemToDelete.getProduct().setStock(itemToDelete.getProduct().getStock() + itemToDelete.getQuantity());
```

Comment:
This will throw NPE in case product with ID = productId does not exist. We either need to use the other approach for this, or throw the custom exception and show the appropriate message on UI.

Reply:
This function is invoked, when user click on 'delete' button near the existing product in cart. ID comes from form (set in like formaction="${pageContext.servletContext.contextPath}/cart/deleteCartItem/${item.product.id}">) Situation, when productToDelete is empty, is not possible.

Reply:
That's not that critical, but, in general, we should not trust user input.
Please check the video: https://www.dropbox.com/s/38jpsob534u3iep/delete-product.mp4?dl=0 - anyone can adjust HTML and pass the other ID in the form.